pkg/storegateway/indexcache/remote.go

```diff
@@ -95,94 +95,77 @@ func (c *RemoteIndexCache) get(ctx context.Context, typ string, key string) ([]b
-// In case of error, it logs and return an empty cache hits map.
-func (c *RemoteIndexCache) FetchMultiPostings(ctx context.Context, userID string, blockID ulid.ULID, lbls []labels.Label) (hits map[labels.Label][]byte, misses []labels.Label) {
-	// Build the cache keys, while keeping a map between input matchers and the cache key
-	// so that we can easily reverse it back after the GetMulti().
-	keys := make([]string, 0, len(lbls))
-	keysMapping := make(map[labels.Label]string, len(lbls))
+// FetchMultiPostings fetches multiple postings - each identified by a label.
+// In case of error, it logs and return an empty result.
+func (c *RemoteIndexCache) FetchMultiPostings(ctx context.Context, userID string, blockID ulid.ULID, lbls []labels.Label) BytesResult {
+	blockIDStr := blockID.String()
 
+	keys := make([]string, 0, len(lbls))
 	for _, lbl := range lbls {
-		key := postingsCacheKey(userID, blockID, lbl)
-
-		keys = append(keys, key)
-		keysMapping[lbl] = key
+		keys = append(keys, postingsCacheKey(userID, blockIDStr, lbl))
 	}
 
 	// Fetch the keys from the remote cache in a single request.
 	c.requests.WithLabelValues(cacheTypePostings).Add(float64(len(keys)))
 	results := c.remote.GetMulti(ctx, keys)
-	if len(results) == 0 {
-		return nil, lbls
-	}
-
-	// Construct the resulting hits map and list of missing keys. We iterate on the input
-	// list of labels to be able to easily create the list of ones in a single iteration.
-	hits = make(map[labels.Label][]byte, len(results))
-	misses = make([]labels.Label, 0, len(keys)-len(results))
-
-	for _, lbl := range lbls {
-		key, ok := keysMapping[lbl]
-		if !ok {
-			level.Error(c.logger).Log("msg", "keys mapping inconsistency found in remote index cache client", "type", "postings", "label", lbl.Name+":"+lbl.Value)
-			continue
-		}
+	c.hits.WithLabelValues(cacheTypePostings).Add(float64(len(results)))
 
-		// Check if the key has been found in the remote cache. If not, we add it to the list
-		// of missing keys.
-		value, ok := results[key]
-		if !ok {
-			misses = append(misses, lbl)
-			continue
-		}
-
-		hits[lbl] = value
+	return &MapIterator[string]{
+		Keys: keys,
+		M:    results,
 	}
-
-	c.hits.WithLabelValues(cacheTypePostings).Add(float64(len(hits)))
-	return hits, misses
 }
 
 // postingsCacheKey returns the cache key used to store postings matching the input
 // label name/value pair in the given block.
-func postingsCacheKey(userID string, blockID ulid.ULID, l labels.Label) string {
+func postingsCacheKey(userID, blockID string, l labels.Label) string {
 	const (
 		prefix    = "P2:"
 		separator = ":"
 	)
 
 	// Compute the label hash.
-	lblHash := postingsCacheKeyLabelHash(l)
+	lblHash, hashLen := postingsCacheKeyLabelID(l)
 
 	// Preallocate the byte slice used to store the cache key.
-	expectedLen := len(prefix) + len(userID) + 1 + ulid.EncodedSize + 1 + base64.RawStdEncoding.EncodedLen(blake2b.Size256)
+	encodedHashLen := base64.RawURLEncoding.EncodedLen(len(lblHash[:hashLen]))
```

Comment:
[nit] Isn't this the same?
```suggestion
	encodedHashLen := base64.RawURLEncoding.EncodedLen(hashLen)
```